client/my-sites/plans-features-main/hooks/use-action-callback.ts

```diff
@@ -0,0 +1,204 @@
+		( params: UseActionCallbackParams, freeTrialPlanSlug: PlanSlug | undefined ) => {
+			return ( isFreeTrialCTA?: boolean ) => {
+				const { planSlug } = params;
+				const upgradePlan = isFreeTrialCTA && freeTrialPlanSlug ? freeTrialPlanSlug : planSlug;
+
+				if ( ! isFreePlan( planSlug ) ) {
+					recordTracksEvent?.( 'calypso_plan_features_upgrade_click', {
+						current_plan: sitePlanSlug,
+						upgrading_to: upgradePlan,
+						saw_free_trial_offer: !! freeTrialPlanSlug,
+					} );
+				} else {
+					recordTracksEvent( 'calypso_signup_free_plan_click' );
+				}
+				processCartItemsForPlanAndAddOns?.( params, freeTrialPlanSlug, isFreeTrialCTA );
+			};
+		},
+		[ sitePlanSlug, processCartItemsForPlanAndAddOns ]
+	);
+}
+
+function useGenerateActionCallback(
+	// TODO: Reevaluate param types and whether or not they should be optional
+	currentPlan: Plans.SitePlan | undefined,
+	cartHandler?: ( cartItems?: MinimalRequestCartProduct[] | null ) => void,
+	flowName?: string | null,
+	freeTrialPlanSlugs?: {
+		[ Type in PlanType ]?: PlanSlug;
+	},
+	intent?: PlansIntent | null,
+	planActionCallback?: ( planSlug: PlanSlug ) => boolean,
+	sitePlanSlug?: PlanSlug | null,
+	siteSlug?: string | null,
+	withDiscount?: string
+): UseActionCallback {
+	const upgradeHandler = useUpgradeHandler(
+		sitePlanSlug,
+		siteSlug,
+		withDiscount,
+		planActionCallback,
+		cartHandler
+	);
+
+	const currentPlanManageHref = useCurrentPlanManageHref();
+	const { managePlan, manageAddOn, gotoVip } = useMemo( () => {
+		function composeActionCallback( callback: () => void ) {
+			return ( { planSlug }: Pick< UseActionCallbackParams, 'planSlug' > ) => {
+				const earlyReturn = planActionCallback?.( planSlug );
+				return earlyReturn ? () => {} : callback;
+			};
+		}
+
+		const managePlan = composeActionCallback( () => page( currentPlanManageHref ) );
+		const manageAddOn = composeActionCallback( () => page.redirect( `/add-ons/${ siteSlug }` ) );
+		const gotoVip = composeActionCallback( () => {
+			recordTracksEvent( 'calypso_plan_step_enterprise_click', { flow: flowName } );
+			window.open( 'https://wpvip.com/wordpress-vip-agile-content-platform', '_blank' );
+		} );
+
+		return { managePlan, manageAddOn, gotoVip };
+	}, [ currentPlanManageHref, flowName, siteSlug ] );
+
+	return (
+		params: UseActionCallbackParams = {
+			planSlug: PLAN_FREE,
+			cartItemForPlan: { product_slug: '' },
+			selectedStorageAddOn: undefined,
+		}
+	) => {
+		const { planSlug } = params;
+		const planConstantObj = applyTestFiltersToPlansList( planSlug, undefined );
+		const freeTrialPlanSlug = freeTrialPlanSlugs?.[ planConstantObj.type ];
+
+		if ( planSlug && isWpcomEnterpriseGridPlan( planSlug ) ) {
+			return gotoVip( params );
+		}
+
+		if ( sitePlanSlug && currentPlan?.productSlug === planSlug && intent !== 'plans-p2' ) {
+			return isFreePlan( sitePlanSlug ) ? manageAddOn( params ) : managePlan( params );
+		}
+
+		return upgradeHandler( params, freeTrialPlanSlug );
```

Comment:
I wonder if handling this a little differently might help avoid `composeActionCallback` and all the named callbacks. Abstractions aren't always beneficial, and I feel in this case there's more complexity being introduced. More so if something changes e.g. no longer sending to "vip"..

Something along these lines:

```
...
const earlyReturn = planActionCallback?.( planSlug );

if ( earlyReturn ) return;

/* 1. Send user to VIP if it's an enterprise plan */
if ( isWpcomEnterpriseGridPlan( planSlug ) ) {
    recordTracksEvent( 'calypso_plan_step_enterprise_click', { flow: flowName } );
    window.open( 'https://wpvip.com/wordpress-vip-agile-content-platform', '_blank' );
    return;
} 

/* 2. Send user to either manage add-ons or plan in case of current plan selection */
if ( currentPlan?.productSlug === planSlug && intent !== 'plans-p2' ) {
    if ( isFreePlan( planSlug ) {
        page.redirect( `/add-ons/${ siteSlug }` );
    } else {
        page.redirect( currentPlanManageHref );
    }
    return;
}

/* 3. Handle a plan upgrade */
handleUpgrade( params, freeTrialPlanSlug );
return;

```

1. No need to call `planActionCallback?.( planSlug )` from `processCartItems` above either. Just here should suffice from the looks of it. or is that not the case? 🤔 
2. renamed `upgradeHandler` to `handleUpgrade`, to be more in line with "some action happening".

Reply:
@chriskmnds After seeing your writeup, I agree. It's noticeably more readable.

> Abstractions aren't always beneficial, and I feel in this case there's more complexity being introduced

I 100% agree with your statement about abstractions, and I'm _so_ curious about the how you intuited that there was complexity. Was it your "spidey sense" again that nudged you in this direction? Was it because it was simply difficult to keep track of the train of logic?

I'm hoping to understand so that I can more clearly recognize it and address it in the future 🙏

Reply:
> No need to call planActionCallback?.( planSlug ) from processCartItems above either. Just here should suffice from the looks of it. or is that not the case? 🤔

It's requiring a modest amount of refactoring because the `planSlug` in `planActionCallback?.( planSlug )` might be a traditional `planSlug` or a `freeTrialPlanSlug` depending on what CTA is clicked at runtime. This _does_ affect the result of the early return. Existing code would mean that we would have to move early return logic back into the `handleUpgrade` function.

I'm making some updates now to accommodate our changes described in https://github.com/Automattic/wp-calypso/pull/88694#discussion_r1561079787, and you can take a look at them tomorrow to see if they make sense

Reply:
https://github.com/Automattic/wp-calypso/pull/88694/commits/2a04d27cc3c0248406a7c275357cc91aab76f18e and https://github.com/Automattic/wp-calypso/pull/88694/commits/9377f1483b037cb2811a389bbd4b9a149efcd9fe

Reply:
> Was it your "spidey sense" again that nudged you in this direction? Or maybe it was simply difficult to keep track of the train of logic?

Not sure. It seemed like something being over-engineered perhaps. A little too generic naming throughout? Generic abstractions are sometimes either incredibly valuable or a signal to the contrary, I think. Especially when names like "compose" or "observer" or "refresh" or "transpose" (etc) pop up - they beg questioning 😁 

(not saying that the abstraction introduced was "bad", just that I found it a little harder to digest than a few conditions and some commenting around)